### PR TITLE
feat(cli): typed v2 harness session-event stream (capture + status + consumer + hooks)

### PR DESCRIPTION
**Pull request by oxoxDev**
## Summary

Adds a **typed v2 session-event stream** (`tinyplace.harness.session.v2`) to the harness bridge, so a consumer (OpenHuman's Master Session) can tell apart **tool_call · agent_message · tool_result · status · everything else** from one field — instead of the v1 text-only stream that only carries "who spoke."

Builds on #229 (merged). Emission is **opt-in and off by default**, so current v1 consumers are byte-identical until they understand v2.

## What's here

- **v2 event model** (`types/harness.ts`) — additive over v1: reuses `bucket`/`scope`/`harness` verbatim, swaps only `message`+`source` for a typed `event`. Two axes: `role` (owner|agent) + `kind` (`user_prompt|agent_message|agent_thinking|approval_request|tool_call|tool_result|status|lifecycle|error|unknown`). v1 and v2 coexist; discriminate on `envelope_version`. `SessionEnvelope` alias stays v1-only so no current importer changes.
- **JSONL→event mapper** (`harness-events.ts`) — splits Claude `tool_use`/`tool_result`/`thinking`/`text` and Codex `function_call`/`function_call_output`/`reasoning`/messages into per-block typed events. `tool_call`↔`tool_result` correlate via `call_id`. Registry dispatch so a 3rd provider (e.g. gemini) is a compile-checked add.
- **Hooks capture path** (`harness-hooks.ts`) — maps Claude Code hook payloads (PreToolUse/PostToolUse/UserPromptSubmit/Stop/SessionStart/…) into the same event shape, for lower latency than JSONL tailing. `tool_use_id` on both PreToolUse and PostToolUse gives exact call↔result correlation.
- **Status state machine** (`harness-status.ts`) — derives `running_tool`/`waiting_approval`/`idle` + heartbeat → drives InstanceStatus + currentTask.
- **Envelope builder** (`harness-envelope.ts`) — seq + stable content-addressed id + bucket window.
- **Consumer** (`harness-consumer.ts`) — `parseSessionEnvelope` + `applySessionEnvelope`/`foldSessionEnvelopes` fold the stream into a live `SessionView` (status, currentTask, tool log, capped feed). Framework-agnostic reference decoder.
- **Wired into the tailer** (`harness-wrapper.ts`) — emits v2 behind `TINYPLACE_HARNESS_V2=1` (default off). Publisher/writeEnvelope accept either version via an `envelope_version` discriminant.

## Testing

- `tsc` build clean; SDK lint (`tsc --noEmit`) clean.
- **442 unit tests pass** (98 staging/network-gated skipped), incl. an integration test that drives the real tailer against a fixture Claude session in dry-run and asserts `tool_call`/`tool_result`/`status` with correlated `call_id` and monotonic seq.
- Provider mapping verified against real Claude + Codex transcripts; hook field names extracted from the Claude Code 2.1.202 binary.

## Notes

- v2 wire shape is the consumer contract. cc @senamakel — this extends your bridge; the v2 envelope reuses your v1 framing so the transport is unchanged.

## Summary by CodeRabbit

* **New Features**
  * Added v2 harness session envelope support with typed event-streams, deterministic event IDs, and safe folding into live session view state (ordered batching, capped feed, tool-call/result + lifecycle/status tracking).
  * Added opt-in v2 emission in the harness wrapper, including derived status updates with heartbeat and idle transitions.
  * Added Claude/Codex JSONL parsing and Claude Code hook mapping into shared semantic events.
* **Public API Updates**
  * Expanded TypeScript and Node exports for v2 envelope contracts plus event parsing, status derivation, and consumer view utilities.
* **Tests**
  * Added Vitest coverage for envelope building, safe parsing/folding edge cases, status derivation, and wrapper v2 emission.

### CODE REVIEW

File: sdk/typescript/src/types/harness.ts [hunk 1]
```diff
@@ -44,3 +44,167 @@ export interface SessionEnvelopeV1 {
...
+export interface AgentThinkingPayload {
+  text: string;
+}
+
+export interface ToolCallPayload {
+  call_id: string;
+  tool_name: string; // raw harness name, e.g. "Bash"
+  tool_kind: HarnessToolKind;
+  display: string; // one-line human summary, e.g. "npm test"
+  input: unknown; // bounded + redacted by the tailer before publish
```
Comment (sanil-23) [inline sdk/typescript/src/types/harness.ts:118]: **Major — contract not honored; unbounded + potential secret exposure.**

This comment promises `input` is `bounded + redacted by the tailer before publish`, but nothing does either. `harness-events.ts:191` (Claude `tool_use`), `:294`/`:343` (Codex `function_call`/`mcp`), and `harness-hooks.ts:138` (`PreToolUse`) all pass the raw `input`/`arguments` straight through; `writeV2` → `buildEventEnvelopeV2` spreads `...semantic.event` and `publishNow` sends `JSON.stringify(envelope)` over the Signal DM with no size guard.

- **Size:** `tool_result.output` is capped at 4KB (`truncate`), but `tool_call.input` is not — a `Write`/`MultiEdit`/`create_file` call carries the entire file, a `Bash` heredoc the entire script, so the call side can dwarf the (carefully bounded) result side and blow past DM limits.
- **Secrets:** tool inputs routinely contain sensitive data (a `Bash` command that `export`s a token, an `Edit` writing a credentials file, an MCP call with an API key). The redaction promise exists for exactly this reason.

Suggest bounding where the payload is built, mirroring the `output` cap:
```ts
// in harness-events.ts, alongside truncate()
function boundInput(input: unknown): unknown {
  const s = safeStringify(input);
  return byteLength(s) <= OUTPUT_CAP ? input : `${s.slice(0, OUTPUT_CAP)}${ELISION}`;
}
// ...then at each pass-through site:
-            input,
+            input: boundInput(input),
```
If secret-key redaction is intended, it belongs here too. At minimum, correct this comment so downstream consumers don't assume a bound that isn't there.

File: sdk/typescript/src/cli/harness-consumer.ts [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+    typeof body === "string" ? tryParse(body) : (body as unknown);
+  if (!record || typeof record !== "object") {
+    return undefined;
+  }
+  const version = (record as { envelope_version?: unknown }).envelope_version;
+  if (
+    version === SESSION_ENVELOPE_VERSION_V2 ||
+    version === SESSION_ENVELOPE_VERSION_V1
+  ) {
+    return record as AnySessionEnvelope;
```
Comment (sanil-23) [inline sdk/typescript/src/cli/harness-consumer.ts:38]: **Major — untrusted-input decoder throws instead of rejecting.**

This module is documented as the reference decoder for DM bodies, and inbound Signal DMs are untrusted. `parseSessionEnvelope` validates only `envelope_version` here and then `return record as AnySessionEnvelope` with no shape check. Downstream, `applySessionEnvelope:115` (`if (event.seq <= view.lastSeq)`) and `foldSessionEnvelopes:143` (`.sort((a, b) => a.event.seq - b.event.seq)`) read `event.seq` unguarded — so a body like `{"envelope_version":"tinyplace.harness.session.v2"}` (no `event`) passes `isV2` and then throws `TypeError: Cannot read properties of undefined (reading 'seq')`, taking down the fold.

Validate the v2 shape before casting:
```ts
if (version === SESSION_ENVELOPE_VERSION_V2) {
  const ev = (record as { event?: { seq?: unknown; id?: unknown } }).event;
  if (!ev || typeof ev.seq !== "number" || typeof ev.id !== "string") return undefined;
  return record as AnySessionEnvelope;
}
if (version === SESSION_ENVELOPE_VERSION_V1) return record as AnySessionEnvelope;
return undefined;
```
Worth a test feeding a malformed v2 body through `parseSessionEnvelope`/`foldSessionEnvelopes` asserting it's dropped, not thrown.

**Review by sanil-23 — CHANGES_REQUESTED** [review]
Nice, cleanly-layered v2 stream with strong test coverage, and the opt-in gating keeps v1 byte-identical — thank you. Two majors to resolve before merge (inline). Everything else in my review is minor/nits:

1. **`tool_call.input` is published verbatim** — unbounded and unredacted — contradicting its own type comment. `tool_result.output` is capped at 4KB but the call-side `input` is not, so a `Write`/`MultiEdit` carries the whole file and a `Bash` heredoc the whole script over the DM, and tool inputs routinely contain secrets (tokens, credential-file writes, MCP keys). Bound it where the payload is built (Claude `harness-events.ts:191`, Codex `:294`/`:343`, hook `harness-hooks.ts:138`) mirroring the `output` cap, and redact if that's what the comment intends — or fix the comment to match reality.

2. **The untrusted-DM decoder crashes on a version-tagged but malformed envelope.** `parseSessionEnvelope` validates only `envelope_version` before casting, and `applySessionEnvelope`/`foldSessionEnvelopes` then read `event.seq` unguarded — a `{"envelope_version":"tinyplace.harness.session.v2"}` body (no `event`) throws instead of being dropped. Validate the v2 shape before the cast.

Minor/nits (not blocking): `truncate` slices by UTF-16 code units after a byte check (~4× loose for multibyte); each semantic event also emits a derived `status` envelope (~2× DM volume); `event.id` is described as content-addressed but hashes `seq`, not payload, and `v2Seq` resets to 0 on restart; `Number(env ?? default)` yields NaN and silently disables idle on non-numeric env. Happy to expand any of these.

File: sdk/typescript/src/index.ts [hunk 3]
```diff
@@ -87,14 +87,74 @@ export {
...
+} from "./cli/harness-events.js";
+export type {
+  HarnessLineMapper,
+  HarnessSemanticEvent,
+} from "./cli/harness-events.js";
+export {
+  claudeHookEventToSemantic,
+  claudeHookEventsFromStdin,
+  claudeHookStringToSemantic,
+} from "./cli/harness-hooks.js";
```
[the remark this answers is not in the document]
Comment (oxoxDev) [reply sdk/typescript/src/index.ts]: Addressed in 698075a7 — moved the harness runtime value exports (events / hooks / status / envelope / consumer) to the `./node` subpath (`@tinyhumansai/tinyplace/node`). The browser-facing root now re-exports only their (type-erased) types, so client bundles no longer resolve `node:process` / `node:crypto`. Verified `dist/node/index.js` carries the moved symbols and the website (which imports none of these values) still typechecks.

File: sdk/typescript/src/cli/harness-events.ts [hunk 4]
```diff
@@ -0,0 +1,539 @@
...
+function truncate(value: string): string {
+  if (byteLength(value) <= OUTPUT_CAP) {
+    return value;
+  }
+  return `${value.slice(0, OUTPUT_CAP)}${ELISION}`;
+}
+
+function byteLength(value: string): number {
+  return Buffer.byteLength(value, "utf8");
+}
```
[the remark this answers is not in the document]
Comment (oxoxDev) [reply sdk/typescript/src/cli/harness-events.ts:496]: Addressed in 5273237a — `truncate()` now slices the UTF-8 `Buffer` (`subarray(0, OUTPUT_CAP)`) instead of `String.slice`, so the byte cap actually holds for multi-byte content; `Buffer.toString("utf8")` drops a trailing partial code point. Added an emoji test asserting the result stays within `OUTPUT_CAP` + elision bytes and contains no replacement char.

File: sdk/typescript/src/cli/harness-hooks.ts [hunk 5]
```diff
@@ -0,0 +1,415 @@
...
+}
+
+/** Keeps small raw payloads structured; truncates large ones to a bounded string. */
+function boundRaw(value: unknown): unknown {
+  const serialized = safeStringify(value);
+  if (byteLength(serialized) <= RAW_CAP) {
+    return value;
+  }
+  return `${serialized.slice(0, RAW_CAP)}${ELISION}`;
+}
```
[the remark this answers is not in the document]
Comment (oxoxDev) [reply sdk/typescript/src/cli/harness-hooks.ts:372]: Addressed in 5273237a — same byte-safe slice applied to both `truncate()` and `boundRaw()` in harness-hooks.ts (the 369 and 387-396 spots).

File: sdk/typescript/src/cli/harness-events.ts [hunk 6]
```diff
@@ -0,0 +1,539 @@
...
+        recordType: "assistant:tool_use",
+        event: {
+          kind: "tool_call",
+          role: "agent",
+          payload: {
+            call_id: asString(object.id) ?? "",
+            tool_name: toolName,
+            tool_kind: normalizeToolKind(toolName),
+            display: toolDisplay(toolName, input),
+            input,
```
[the remark this answers is not in the document]
Comment (oxoxDev) [reply sdk/typescript/src/cli/harness-events.ts]: Addressed in 70b64369 — raw `tool_input` is now byte-capped (`INPUT_CAP` = 4096) via `boundToolInput()` before it enters the `tool_call` envelope, across the Claude, Codex and hook mappers; small inputs keep their object structure. Added a test with an oversized Write input asserting the published `input` is a byte-bounded string.

File: sdk/typescript/src/cli/harness-events.ts [hunk 7]
```diff
@@ -0,0 +1,539 @@
...
+    return undefined;
+  }
+}
+
+function parseTimestamp(value: unknown): Date {
+  if (typeof value !== "string") {
+    return new Date(0);
+  }
+  const parsed = new Date(value);
+  return Number.isNaN(parsed.getTime()) ? new Date(0) : parsed;
```
[the remark this answers is not in the document]
Comment (oxoxDev) [reply sdk/typescript/src/cli/harness-events.ts]: Addressed in 5e36a026 — `parseTimestamp` now defaults missing/invalid timestamps to receive time (`new Date()`) instead of the epoch, so `reduceStatus`'s activity clock advances and `tickStatus` no longer idles a live session right after a tool call. Added a timestamp-less Codex `function_call` regression test asserting the status stays `running_tool` and doesn't emit idle on the next tick.

File: sdk/typescript/src/cli/harness-events.ts [hunk 8]
```diff
@@ -0,0 +1,539 @@
...
+            call_id: asString(payload.call_id) ?? asString(payload.id) ?? "",
+            tool_name: toolName,
+            tool_kind: normalizeToolKind(toolName),
+            display: toolDisplay(toolName, input),
+            input,
+          },
+        },
+      },
+    ];
+  }
```
[the remark this answers is not in the document]
Comment (oxoxDev) [reply sdk/typescript/src/cli/harness-events.ts:301]: Addressed in f6b84c59 — Codex `function_call` / `tool_search_call` / `mcp_tool_call_begin` `arguments` (a JSON string) are now parsed via `parseMaybeJson` before deriving `display` and `input`, so the shape matches Claude's object form. Added a test asserting the parsed `input` object and a field-derived `display`.

File: sdk/typescript/src/cli/harness-status.ts [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+    case "turn_start":
+      return { state: "running", detail: "working" };
+    case "turn_end":
+      return { state: "idle", detail: "idle" };
+    case "session_end":
+      return { state: "stopped", detail: "stopped" };
+    default:
+      return undefined;
+  }
+}
```
[the remark this answers is not in the document]
Comment (oxoxDev) [reply sdk/typescript/src/cli/harness-status.ts:158]: Addressed in 8ecc25fe — lifecycleStatus now handles the `compact` phase explicitly, surfacing it as `{ state: "running", detail: "compacting" }` instead of falling through the default and leaving a stale status. Added a reducer test for it.

File: sdk/typescript/tests/harness-wrapper-v2.test.ts [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+  const tailer = new HarnessSessionTailer(config, "/work/proj", out, publisher);
+  tailer.start(new Date("2026-07-07T09:59:00.000Z"));
+  await tailer.stop();
+}
+
+describe("HarnessSessionTailer v2 emission", () => {
+  const dirs: Array<string> = [];
+  afterEach(() => {
+    dirs.length = 0;
+  });
```
[the remark this answers is not in the document]
Comment (oxoxDev) [reply sdk/typescript/tests/harness-wrapper-v2.test.ts:108]: Addressed in 770288fc — afterEach now `rmSync(dir, { recursive: true, force: true })` on each tracked mkdtemp dir before clearing the array.

File: sdk/typescript/src/cli/harness-envelope.ts [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+): SessionEnvelopeV2 {
+  const at = usableDate(semantic.timestamp);
+  const start = floorBucket(at, ctx.bucketUnit);
+  const end = addBucket(start, ctx.bucketUnit);
+  const id = stableEventId(
+    ctx.wrapperSessionId,
+    semantic.recordType,
+    seq,
+    semantic.event.kind,
+  );
```
[the remark this answers is not in the document]
Comment (oxoxDev) [reply sdk/typescript/src/cli/harness-envelope.ts:50]: Thanks — I looked at this and I'm keeping `seq` as the id input here, for two reasons:

1. Switching to `semantic.line` would collide. A single transcript line can expand into multiple semantic events — e.g. one Claude `assistant` record with several `content` blocks yields multiple `tool_call` events that share the same `line`, `recordType` and `kind`. `sha256(wrapperSessionId, recordType, line, kind)` would then be identical for those distinct events, so dedup would drop real tool calls. `seq` is what keeps them apart.

2. The computed `event.id` travels on the envelope, so a network-level resend carries the same id and dedups correctly — the id isn't recomputed on resend. `seq` is the documented monotonic-per-session ordering key; id/ts are derived once at emit.

A restart-stable AND unique id would need `line` plus an intra-line block index (or a content hash), which is a broader contract change. Happy to open a follow-up if you want that stronger cross-restart guarantee, but the `line`-only swap would regress correctness so I'd rather not take it as-is.

File: sdk/typescript/src/cli/harness-consumer.ts (hunk 2 above)
Comment (sanil-23) [inline sdk/typescript/src/cli/harness-consumer.ts:38]: **Still open (Major #2).** Unchanged since the last review — this casts on `envelope_version` alone, and `applySessionEnvelope`/`foldSessionEnvelopes` then read `event.seq` unguarded, so a version-tagged body with no `event` throws `TypeError: Cannot read properties of undefined (reading 'seq')` rather than being dropped. Given the module decodes untrusted inbound DMs, validate the v2 shape before casting:
```ts
if (version === SESSION_ENVELOPE_VERSION_V2) {
  const ev = (record as { event?: { seq?: unknown; id?: unknown } }).event;
  if (!ev || typeof ev.seq !== "number" || typeof ev.id !== "string") return undefined;
  return record as AnySessionEnvelope;
}
if (version === SESSION_ENVELOPE_VERSION_V1) return record as AnySessionEnvelope;
return undefined;
```
Plus a test feeding a malformed v2 body through `parseSessionEnvelope`/`foldSessionEnvelopes` asserting it's dropped, not thrown.

**Review by sanil-23 — COMMENTED** [review]
Re-reviewed at `770288fc` — thanks for the fast turnaround. **Major #1 (unbounded/unredacted `tool_call.input`) is resolved**: `boundToolInput`/`boundRaw` now byte-cap input at every emit site (Claude `tool_use`, Codex `function_call` + MCP, hook `PreToolUse`), structure preserved when small. You also picked up all my minors and then some — byte-indexed `truncate` in both mappers, `parseMaybeJson` on Codex arguments so `input`/`toolDisplay` match Claude's object shape, receive-time fallback for timestamp-less events (fixes the premature-idle edge), the `compact` lifecycle phase in the status reducer, and the Node-only `/node` subpath split to keep browser bundles clean. All covered by new tests. 👍

**One item still open — Major #2:** the consumer still crashes on a malformed-but-version-tagged DM body. `parseSessionEnvelope` (`harness-consumer.ts:34`) validates only `envelope_version` before casting, and `applySessionEnvelope:115` / `foldSessionEnvelopes:143` read `event.seq` unguarded — so `{"envelope_version":"tinyplace.harness.session.v2"}` still throws instead of being dropped. Since this is the reference decoder for untrusted inbound DMs, worth shape-validating before the cast (fix + suggested test in the inline comment). Once that lands I'm happy to approve.

File: sdk/typescript/src/cli/harness-consumer.ts (hunk 2 above)
Comment (oxoxDev) [reply sdk/typescript/src/cli/harness-consumer.ts:38]: Addressed in c93fc44a — parseSessionEnvelope now shape-checks the v2 event (numeric seq + string id) via a shared `hasUsableV2Event` guard, wired into both fold paths (applySessionEnvelope + foldSessionEnvelopes .filter) so a malformed-but-version-tagged body is dropped, not thrown — whether it arrives via the parser or pre-parsed. Added 3 tests: parse drops it, apply returns the same view (no throw), fold skips it and still folds the good stream. Full suite 450 pass.

Comment (oxoxDev) [reply sdk/typescript/src/cli/harness-consumer.ts:38]: Fixed in c93fc44a — see the reply on the re-review thread. Shape-validated before cast + guarded both fold reads, with regression tests.

File: sdk/typescript/src/types/harness.ts (hunk 1 above)
Comment (oxoxDev) [reply sdk/typescript/src/types/harness.ts:118]: Confirmed resolved — `boundToolInput`/`boundRaw` byte-cap tool input at every emit site (Claude tool_use, Codex function_call + MCP, hook PreToolUse) and the type comment now matches. Thanks for the catch.

**Review by sanil-23 — APPROVED** [review]
Approving — both blocking findings are resolved and CI is green (Test/Typecheck/Lint/Build all pass; only the unrelated Vercel team-auth check is red).

**Major #2 (malformed-envelope crash) — ✅ fixed, thoroughly.** `hasUsableV2Event` shape-checks `event.seq`/`event.id`, and you gated it in all three paths — `parseSessionEnvelope`, `applySessionEnvelope`, and `foldSessionEnvelopes` — so a poisoned body is dropped whether it arrives raw or pre-parsed. Tests cover missing-event, missing-seq, non-string-id, and the poisoned-stream-doesn't-throw case. That's better than the minimal fix I suggested.

**Major #1 (unbounded/unredacted `tool_call.input`) — ✅ fixed** last round (`boundToolInput`/`boundRaw` at every emit site).

Recap of everything addressed across the three rounds: input bounding, byte-indexed truncation, Codex `parseMaybeJson`, receive-time timestamp fallback, `compact` lifecycle mapping, the `/node` subpath split to keep Node built-ins out of browser bundles, and now the untrusted-decoder hardening. Nice work — the v2 stream is clean, well-typed, well-tested, and opt-in. LGTM. 🚀

**Comment by oxoxDev** [comment]
@sanil-23 Major #2 is fixed (`c93fc44a`), and I swept the four minors from your review body too:

1. **`truncate` sliced by UTF-16 after a byte check** — fixed in `5273237a`: both mappers now byte-truncate via `Buffer.subarray` (drops a trailing partial code point), with a multibyte regression test.
2. **Derived `status` doubling DM volume** — the status emit is *change-gated*: `reduceStatus` returns an envelope only when the derived state/detail actually changes (idle→running_tool→running→idle), not once per content event, so it's bounded to state transitions rather than 2× the stream. Happy to coalesce status into the next content envelope if you'd still prefer fewer DMs.
3. **`event.id` described as content-addressed but keyed on `seq`, resets on restart** — clarified in `1704e3e3`. It's deterministic over `(wrapperSessionId, recordType, seq, kind)`, not a payload hash: stable within a run for resend-dedup, and collision-free across runs because `wrapperSessionId` is minted per process — so `seq` restarting at 0 after a relaunch can't collide with a prior run's ids. (Corrected the PR-description wording too.)
4. **`Number(env ?? default)` → NaN** — fixed in `7745b9e7`: all five status/poll knobs route through `numberEnvOr`, which falls back to the default on unset *or* non-finite input, so a typo'd env var can't silently disable the idle/heartbeat tick.

Full suite 450 pass, build clean. Re-requested your review — thanks for the thorough pass.

File: sdk/typescript/src/cli/harness-wrapper.ts [hunk 12]
```diff
@@ -415,33 +438,55 @@ export function parseHarnessWrapperArgs(
...
         "TINYPLACE_HARNESS_SESSIONS_DIR",
       ]) ?? profile.defaultSessionsDir,
-    sessionTailGraceMs: Number(
+    sessionTailGraceMs: numberEnvOr(
       firstEnv(env, [
         `TINYPLACE_${provider.toUpperCase()}_SESSION_TAIL_GRACE_MS`,
         "TINYPLACE_HARNESS_SESSION_TAIL_GRACE_MS",
-      ]) ?? 750,
+      ]),
+      750,
```
[the remark this answers is not in the document]
Comment (oxoxDev) [reply sdk/typescript/src/cli/harness-wrapper.ts:489]: Addressed in c27c1452 — numberEnvOr now trims and requires a finite value > 0, so unset / non-numeric / blank (Number("")→0) / negative all fall back to the default instead of poisoning the timer (disabled tick or 0 ms poll loop). Build + wrapper tests green.

**Comment by oxoxDev** [comment]
@coderabbitai the one actionable from the last pass — the over-permissive `numberEnvOr` (blank→0 / negative timers) — is fixed in `c27c1452`: it now trims and requires a finite value > 0, so unset, non-numeric, blank, and negative env values all fall back to the default instead of disabling the idle/heartbeat tick or spinning a 0 ms poll loop. All prior threads (byte-safe truncation in events + hooks, tool_input bounding, Codex timestamp/arguments handling, consumer shape-guard, temp-dir cleanup) remain addressed. Build + tests green. Please re-review.
